### PR TITLE
Drawer + Tabs

### PR DESCRIPTION
**Pull request by asapMaki**
## Description
First version of Drawer + Tabs for Expo router

## Motivation and Context
To have the ability to choose Drawer + Tabs as an option to the CLI

## How Has This Been Tested?
Locally

### CODE REVIEW

File: cli/package.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
     "clean-build": "rm -rf ./build",
     "compile": "tsc -p .",
     "copy-templates": "bun run copyfiles -u 2 -a \"./src/templates/**/*\" ./build/templates",
-    "format": "eslint \"**/*.{js,jsx,ts,tsx}\" --fix && prettier \"**/*.{js,jsx,ts,tsx,json}\" --write",
+    "format": "eslint \"**/*.{js,jsx,ts,tsx}\" --fix --resolve-plugins-relative-to . && prettier \"**/*.{js,jsx,ts,tsx,json}\" --write",
```
Comment (danstepanov) [inline cli/package.json:29]: why was this added?

File: cli/src/templates/packages/expo-router/drawer/app/_layout.tsx.ejs [hunk 2]
```diff
@@ -1,47 +1,48 @@
...
 import { GestureHandlerRootView } from 'react-native-gesture-handler';
 import { Slot } from 'expo-router';
-<% if (props.stylingPackage?.name === "tamagui") { %>
-	import React, { useEffect } from "react";
-	import { TamaguiProvider } from 'tamagui'
-	import { SplashScreen } from "expo-router";
-	import { useFonts } from "expo-font";
 
-	import config from '../tamagui.config'
+<% if (props.stylingPackage?.name==="tamagui" ) { %>
```
Comment (danstepanov) [inline cli/src/templates/packages/expo-router/drawer/app/_layout.tsx.ejs:5]: all changes in this file seem to be code formatting?

File: cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
```
Comment (danstepanov) [inline cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs:1]: What are these files for?

File: cli/package.json (hunk 1 above)
Comment (asapMaki) [reply cli/package.json:29]: image

So running `yarn create-expo-stack` has found multiple prettier instances
So this enables it to choose the one relative to the folder I am in.
This happens to me as I am running this command inside of current cli to do fast iteration of changes.

image

File: cli/src/templates/packages/expo-router/drawer/app/_layout.tsx.ejs (hunk 2 above)
Comment (asapMaki) [reply cli/src/templates/packages/expo-router/drawer/app/_layout.tsx.ejs:5]: It really does seems like it, I was trying to setup my env and see how formatting was done.
Prettier didn't format it at all on save or manually, nor other formatters.
Can you share maybe extension you are using ?

This file was handled by copilot.
And it seems like it was following the configs correctly, especially the tab width ?
Screenshot 2024-01-13 at 08 32 40

File: cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs (hunk 3 above)
Comment (asapMaki) [reply cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs:1]: Minimal screens for bottom tabs

File: cli/src/templates/packages/expo-router/drawer/app/_layout.tsx.ejs (hunk 2 above)
Comment (danstepanov) [reply cli/src/templates/packages/expo-router/drawer/app/_layout.tsx.ejs:5]: interesting, I don't think these formatting instructions are being applied to .ejs files. I know it might be a bit annoying and we're working on better tools to work with ejs files but can you try to manually change the formatting stuff back? It will make it much easier to attribute your changes to actual functionality of this PR.

File: cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs (hunk 3 above)
Comment (danstepanov) [reply cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs:1]: is there a reason that you chose to create separate files for these components vs just declaring them inline?

File: cli/package.json (hunk 1 above)
Comment (danstepanov) [reply cli/package.json:29]: @ernestoresende @sammoore thoughts?

File: cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs (hunk 3 above)
Comment (asapMaki) [reply cli/src/templates/packages/expo-router/drawer/app/(drawer)/(tabs)/TabTwo.tsx.ejs:1]: Usually React Navigation leaves warning for inline screens, so it is probably habit.
But also easier for people to copy paste file of screen and start changing.

Most of these changes in PR are bare minimum for you guys to review my approach to this CLI.

**Comment by danstepanov** [comment]
@asapMaki can you rebase this PR with `main`? Recently merged a large PR from @ludwig-pro so perhaps he can offer some input if you get lost?

File: cli/package.json (hunk 1 above)
Comment (sammoore) [reply cli/package.json:29]: @danstepanov This is great!

Comment (sammoore) [inline cli/package.json:29]: This is a fantastic change, thanks for the fix!

File: cli/src/templates/packages/react-navigation/navigation/tab-navigator.tsx.ejs [hunk 4]
```diff
@@ -37,7 +41,8 @@ export default function TabLayout({ navigation }) {
                 />
               )}
             </Pressable>
-          ),
+            )
```
Comment (sammoore) [inline cli/src/templates/packages/react-navigation/navigation/tab-navigator.tsx.ejs:44]: I know Dan said above to remove the formatting changes in the ejs files, but while you're at it, we might as well fix the indentation here ;)

**Review by sammoore — CHANGES_REQUESTED** [review]
This mostly looks great, except for the comments I made, and that there are some patches/changeset file changes I'd recommend:

- There are a few duplicate patches with the description `Removed formatting flag from package json command` -- I think you've since changed it to `--resolve-config-relative-to .`, but also that sort of change is not necessary to mention in the changelog, so I would remove all of those.
- I think the two other patches related to the "Drawer + Tabs" thing could be merged into one changeset, so we're not mentioning the Drawer + Tabs twice in the changelog after this is merged cc @danstepanov

**Comment by danstepanov** [comment]
This PR is missing the following:
- [ ] Ability to generate a drawer+tabs project via non-interactive mode (ie pass in `--drawer+tabs`
- [ ] Proper output of `generateRerunScript`
- [ ] Correct back button in react-navigation template in the Modal presentation
Screenshot 2024-01-25 at 3 37 37 AM
- [ ] Type errors in `drawer-navigator.tsx` in the react-navigation template

**Comment by danstepanov** [comment]
Completed 4/5 tasks, last one is relatively minor and I don't know enough about React Navigation to fix. I created https://github.com/danstepanov/create-expo-stack/issues/191 to track the issue and we can address in a future release.
